### PR TITLE
Add e2e test framework and some e2e tests

### PR DESCRIPTION
**Pull request by nils-wisiol**
**Note**: This PR is currently set up against the branch of PR #70, but should eventually be merged into master.

This PR adds an e2e test framework, configures travis to run e2e and api tests (including some time and handling improvements), extends the e2e test framework to be able to check DNS responses and finally adds sample tests for functions of the `www` container and of the dyndns12 endpoint of the `api` container.

More e2e tests should be added, especially to verify functionality of the `rrset` endpoint.

### CODE REVIEW

File: api/wait [hunk 1]
```diff
@@ -0,0 +1,8 @@
+#!/usr/bin/env bash
```
Comment (peterthomassen) [inline api/wait]: Maybe this file should be named `wait.sh`, not have a shebang line and instead host a function `wait_for_deps` which you can call from the entrypoint scripts (after sourcing). This would look a bit cleaner, and standalone execution of this library-like functionality would not be possible.

File: docker-compose.test-e2e.yml [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    - DESECSTACK_IPV6_SUBNET
+    - DESECSTACK_IPV6_ADDRESS
+    mac_address: 06:42:ac:10:00:7f
+    depends_on:
+    - www
+    - nslord
+    networks:
+      front:
+        ipv4_address: ${DESECSTACK_IPV4_REAR_PREFIX16}.0.127
+    extra_hosts:
```
Comment (peterthomassen) [inline docker-compose.test-e2e.yml]: cute!

File: test/e2e/apiwait [hunk 3]
```diff
@@ -0,0 +1,22 @@
+#!/usr/bin/env bash
+
+if [ -f ./.env ] ; then
+    source ../../.env
+fi
+
+TIME=0
+LIMIT=${1:-3}
```
Comment (peterthomassen) [inline test/e2e/apiwait]: very intuitive, that `-` sign :-D

File: test/e2e/spec/test_configuration_spec.js [hunk 4]
```diff
@@ -0,0 +1,18 @@
+var chakram = require('./../setup.js').chakram;
+var expect = chakram.expect;
+
+describe("test configuration", function () {
+
+    it("has a hostname", function () {
+        return expect(process.env.BOILERPLATE_DOMAIN).to.not.equal('');
```
Comment (peterthomassen) [inline test/e2e/spec/test_configuration_spec.js]: Do we have `BOILERPLATE_` variables here? (I thought they're called `DESECSTACK_`)

File: test/e2e/spec/www_spec.js [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+            it('returns an ipv6 address when contacted through ipv6', function () {
+                var response = chakram.get('https://[' + process.env.DESECSTACK_IPV6_ADDRESS + ']/');
+
+                // it's hard to find out which IPv6 address we actually expect here
+                // and as we are inside the docker network anyway (that is, we are
+                // topologically not in the same place as the end user), it's hard
+                // if the correct address is returned. Hence, we will stick to some
+                // simple tests.
+                expect(response).to.have.body(REGEX_IPV6_ADDRESS);
+                expect(response).to.not.have.body(process.env.DESECSTACK_IPV4_REAR_PREFIX16 + '.0.127');
```
Comment (peterthomassen) [inline test/e2e/spec/www_spec.js]: don't these two (positive) checks exclude each other anyways?

File: test/e2e/spec/www_spec.js [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+
+            // it's hard to find out which IPv6 address we actually expect here
+            // and as we are inside the docker network anyway (that is, we are
+            // topologically not in the same place as the end user), it's hard
+            // if the correct address is returned. Hence, we will stick to some
+            // simple tests.
+            expect(response).to.have.body(/(([0-9a-fA-F]{1,4}:){7,7}[0-9a-fA-F]{1,4}|([0-9a-fA-F]{1,4}:){1,7}:|([0-9a-fA-F]{1,4}:){1,6}:[0-9a-fA-F]{1,4}|([0-9a-fA-F]{1,4}:){1,5}(:[0-9a-fA-F]{1,4}){1,2}|([0-9a-fA-F]{1,4}:){1,4}(:[0-9a-fA-F]{1,4}){1,3}|([0-9a-fA-F]{1,4}:){1,3}(:[0-9a-fA-F]{1,4}){1,4}|([0-9a-fA-F]{1,4}:){1,2}(:[0-9a-fA-F]{1,4}){1,5}|[0-9a-fA-F]{1,4}:((:[0-9a-fA-F]{1,4}){1,6})|:((:[0-9a-fA-F]{1,4}){1,7}|:)|fe80:(:[0-9a-fA-F]{0,4}){0,4}%[0-9a-zA-Z]{1,}|::(ffff(:0{1,4}){0,1}:){0,1}((25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])\.){3,3}(25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])|([0-9a-fA-F]{1,4}:){1,4}:((25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])\.){3,3}(25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9]))/);
+            expect(response).to.not.have.body(process.env.DESECSTACK_IPV4_REAR_PREFIX16 + '.0.127');
```
Comment (peterthomassen) [inline test/e2e/spec/www_spec.js]: see above

File: api/wait (hunk 1 above)
Comment (nils-wisiol) [reply api/wait]: What's the benefit of doing so?

In a similar case, the `apiwait` script in the e2e test container is called directly from `.travis.yml`. In that case, your proposed changes would yield to additional refactoring there.

File: test/e2e/apiwait (hunk 3 above)
Comment (nils-wisiol) [reply test/e2e/apiwait]: I know. I added a comment...

File: test/e2e/spec/test_configuration_spec.js (hunk 4 above)
Comment (nils-wisiol) [reply test/e2e/spec/test_configuration_spec.js]: fixed (also fixed the reason why those tests didn't fail)

File: test/e2e/spec/www_spec.js (hunk 5 above)
Comment (nils-wisiol) [reply test/e2e/spec/www_spec.js]: yup. fixed

File: test/e2e/spec/www_spec.js (hunk 6 above)
Comment (nils-wisiol) [reply test/e2e/spec/www_spec.js]: fixed

**Comment by nils-wisiol** [comment]
I added a couple of fixes for things you mentions (labeled 'review-fix') and two more things that I noticed along the way.

[One comment is still unresolved](https://github.com/desec-io/desec-stack/pull/71#discussion_r137943953) and needs your feedback.

File: api/wait (hunk 1 above)
Comment (peterthomassen) [reply api/wait]: I don't like having an executable for what's really a library function, but I don't feel strongly. It's ok the way it is.

File: test/e2e/apiwait [hunk 7]
```diff
@@ -5,7 +5,7 @@ if [ -f ./.env ] ; then
 fi
 
 TIME=0
-LIMIT=${1:-3}
+LIMIT=${1:-3}  # getting limit or default to 3 [sic]ter
```
Comment (peterthomassen) [inline test/e2e/apiwait]: `ter`?

Comment (nils-wisiol) [reply test/e2e/apiwait]: fixed

**Comment by nils-wisiol** [comment]
squashed & rebased, can be merged after #70 (don't forget to change the PR base)

File: test/e2e/setup.js [hunk 8]
```diff
@@ -1,6 +1,44 @@
...
 var chakram = require('/usr/local/lib/node_modules/chakram/lib/chakram.js');
+var Q = require('q');
+
+// FIXME contacting nslord for DNS responses. This can changed to nsmaster as soon as changes to the DNS are applied
+// immediately, i.e. before pdns HTTP responses return to the API.
+const dns = require('dns');
+const resolver = new dns.Resolver();
+resolver.setServers([process.env.DESECSTACK_IPV4_REAR_PREFIX16 + '.0.129']);
+
+// Shamelessly taken from https://stackoverflow.com/a/14853974
```
Comment (peterthomassen) [inline test/e2e/setup.js]: remove override, provided tests still run

File: test/e2e/spec/dyndns_spec.js [hunk 9]
```diff
@@ -0,0 +1,289 @@
...
+                    expect(loginResponse.body.auth_token).to.match(/^[a-z0-9]{40}$/);
+                    token = loginResponse.body.auth_token;
+                    chakram.setRequestHeader('Authorization', 'Token ' + token);
+                });
+            });
+        });
+
+        describe("(logged in)", function () {
+
+            before(function () {
```
Comment (peterthomassen) [inline test/e2e/spec/dyndns_spec.js]: remove

File: test/e2e/spec/dyndns_spec.js [hunk 10]
```diff
@@ -0,0 +1,289 @@
...
+
+        var domain = 'e2etest-' + require("uuid").v4() + '.dedyn.io';
+        describe("and domain [" + domain + "]", function () {
+
+            before(function () {
+                chakram.setRequestHeader('Authorization', 'Token ' + token);
+                return expect(chakram.post('/domains/', {'name': domain})).to.have.status(201);
+            });
+
+            describe("dyndns12 endpoint with basic auth", function () {
```
Comment (peterthomassen) [inline test/e2e/spec/dyndns_spec.js]: add tests with empty IPv4 and/or IPv6 argument and test record deletion

Comment (nils-wisiol) [reply test/e2e/spec/dyndns_spec.js]: done:
1) https://github.com/desec-io/desec-stack/pull/71/files#diff-630fb43ba6d02fbc6b856ae12276a939R230
2) https://github.com/desec-io/desec-stack/pull/71/files#diff-630fb43ba6d02fbc6b856ae12276a939R279

**Comment by peterthomassen** [comment]
lgtm, thanks!

Do I understand correctly that this should be merged now, and we add in the `20171208_extend_e2e` branch later?
